constraints/cpu.md

```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
 400000  # cfs quota
 ```
 
+
+## CPU cgroups
+
+Kata Containers run over two layers of cgroups, the first layer is in the guest where
+only the workload is placed, the second layer is in the host that is more complex and
+might contain more than one process and task (thread) depending of the number of
+containers per POD and vCPUs per container. Following diagram represents a nginx container
+created with docker with the default number of vcpus.
+
+
+```
+docker run -dt --runtime=kata-runtime nginx
+
+
+       .-------.
+       | nginx |
+    .--'-------'---.  .------------.
+    | Guest Cgroup |  | Kata agent |
+  .-'--------------'--'------------'.    .-----------.
+  |  Thread: Hypervisor's vCPU 0    |    | Kata Shim |
+ .'---------------------------------'.  .'-----------'.
+ |             Tasks                 |  |  Processes  |
+.'-----------------------------------'--'-------------'.
+|                    Host Cgroup                       |
+'------------------------------------------------------'
+```
+
+In next sections is explained what is the difference between processes and tasks and why
+only hypervisor's vCPUs are constrained.
+
+### cpu cgroups in the guess
+
+Only the workload process including all its threads are placed into cpu cgroups, this means
+that kata agent and systemd run without constraints in the guest.
+
+#### CPU pinning (cpuset cgroup)
+
+Kata Containers try to apply and honor the cgroups but sometimes that's not possible, this
+is the case of cpuset cgroup where the number of virtual CPUs (guest) does not match with
+the actual number of physical CPUs (host).
+In Kata Containers to have a good performance and small memory footprint, the resources are
+hot added when they are needed, therefore the number of virtual resources is not the same
+than the number of physical resources. The problem with this approach is that it's not possible
+to pin a process on a specific resource that is not present in the guest. To deal with this
+limitation and do not fail when the container is being created, Kata Containers do not apply
+the constraint in the first layer (guest) if the resource does not exist in the guest, but it
```

Comment:
s/does not exist/do not exist/

Reply:
probably I'm wrong, but `the resource` is an `it` and it's singular, isn't it?

Reply:
Oops! Good spot - my mistake. Ignore that comment ;)